In [126]:
#데이터 가져오기
import pandas as pd
import numpy as np

data1 = pd.read_csv('/Users/adrua/Desktop/Fariy of Choice/Fairy_of_Choice/resultdata1-1.csv')
data2 = pd.read_csv('/Users/adrua/Desktop/Fariy of Choice/Fairy_of_Choice/resultdata1-2.csv')
data3 = pd.read_csv('/Users/adrua/Desktop/Fariy of Choice/Fairy_of_Choice/resultdata2.csv')
data4 = pd.read_csv('/Users/adrua/Desktop/Fariy of Choice/Fairy_of_Choice/resultdata3.csv')
data5 = pd.read_csv('/Users/adrua/Desktop/Fariy of Choice/Fairy_of_Choice/resultdata4.csv')

#data1.head(3)

In [324]:
#5개 파일로 분리되어 있는 데이터를 하나의 변수에 OUTER JOIN해서 저장 
#import pandas as pd (already done)

merge1 = pd.merge(data1, data2, on = 'userID', how = 'outer')
merge2 = pd.merge(merge1, data3, on = 'userID', how = 'outer')
merge3 = pd.merge(merge2, data4, on = 'userID', how = 'outer')
rating_matrix = pd.merge(merge2, data5, on = 'userID', how = 'outer')
rating_matrix = rating_matrix.fillna(0)
rating_matrix.head(3)
#print(rating_matrix.shape) #>>> 17747행, 2316열

userID  #술비 부평점  1인고기도시락혼자고기야  24시시골집밥상  60계 서울신림난곡점  777카페  88옛날통닭 창동점  \
0  1006bu2      0.0           0.0       0.0          0.0    0.0         0.0   
1  100dlpu      0.0           0.0       0.0          0.0    0.0         0.0   
2  100i22n      0.0           0.0       0.0          0.0    0.0         0.0   

   BBQ 난곡점  BBQ 봉천우성점  BBQ 사당역점  ...  함지박치즈등갈비 신림2호점  핫도리탕  햇빛식당  혼맛스시  \
0      0.0        0.0       0.0  ...             0.0   0.0   0.0   0.0   
1      0.0        0.0       0.0  ...             0.0   0.0   0.0   0.0   
2      0.0        0.0       0.0  ...             0.0   0.0   0.0   0.0   

   홍마떡 신림점  황솔촌 충장점  흥부족발보쌈  흥부찜닭 공수간 삼겹본능 신림점  희래등 호계본점  희래원  
0      0.0      0.0     0.0                0.0       0.0  0.0  
1      0.0      0.0     0.0                0.0       0.0  0.0  
2      0.0      0.0     0.0                0.0       0.0  0.0  

[3 rows x 2316 columns]

In [332]:
#userID·storeID 인덱스 설정
ratings_matrix = rating_matrix.set_index("userID")
ratings_matrix.columns.names = ["storeID"]
ratings_matrix = ratings_matrix.drop(columns=['Unnamed: 482'], axis=1)
ratings_matrix.head(3)
#ratings_matrix[17740:17749]

storeID  #술비 부평점  1인고기도시락혼자고기야  24시시골집밥상  60계 서울신림난곡점  777카페  88옛날통닭 창동점  \
userID                                                                     
1006bu2      0.0           0.0       0.0          0.0    0.0         0.0   
100dlpu      0.0           0.0       0.0          0.0    0.0         0.0   
100i22n      0.0           0.0       0.0          0.0    0.0         0.0   

storeID  BBQ 난곡점  BBQ 봉천우성점  BBQ 사당역점  BBQ 서울대글로벌공학관점  ...  함지박치즈등갈비 신림2호점  \
userID                                                 ...                   
1006bu2      0.0        0.0       0.0             0.0  ...             0.0   
100dlpu      0.0        0.0       0.0             0.0  ...             0.0   
100i22n      0.0        0.0       0.0             0.0  ...             0.0   

storeID  핫도리탕  햇빛식당  혼맛스시  홍마떡 신림점  황솔촌 충장점  흥부족발보쌈  흥부찜닭 공수간 삼겹본능 신림점  \
userID                                                                   
1006bu2   0.0   0.0   0.0      0.0      0.0     0.0                0.0   
100dlpu   0.0   0.0   0.0      0.0      0.0     0.0                0.0   
100i22n   0.0   0.0   0.0      0.0      0.0     0.0                0.0   

storeID  희래등 호계본점  희래원  
userID                  
1006bu2       0.0  0.0  
100dlpu       0.0  0.0  
100i22n       0.0  0.0  

[3 rows x 2314 columns]

In [333]:
#매장별 평균 평점 & 정렬 
ratings_mean = ratings_matrix.replace(0, np.NaN)

ratings_matrix.loc[17748] = ratings_mean.mean(axis = 'rows')
#ratings_matrix.rename(index ={17748:'average'}, inplace = True)
ratings_matrix = ratings_matrix.rename(index ={17748:'average'}, inplace = False)
ratings_matrix = ratings_matrix.sort_values(by='average', axis=1, ascending=False)

ratings_matrix[17740:17749]

storeID  #술비 부평점  상록수온누리약국  서울닭집  서울닭튀김  서울삼족탕  소양강민물장어  손문대구막창갈매기살 시립대점  \
userID                                                                     
vu7pe0       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vuvv60       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vv3ge2       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vv3ucg       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vv3vh1       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vvetso       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
vvm4sd       0.0       0.0   0.0    0.0    0.0      0.0              0.0   
average      5.0       5.0   5.0    5.0    5.0      5.0              5.0   

storeID  수학에미친사람들  술냥이  숯먹은생고기촌  ...  대만족 봉천역점  소소떡볶이 관악점  손큰원조할매순대국 미아점  \
userID                           ...                                       
vu7pe0        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vuvv60        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vv3ge2        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vv3ucg        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vv3vh1        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vvetso        0.0  0.0      0.0  ...       0.0        0.0            0.0   
vvm4sd        0.0  0.0      0.0  ...       0.0        0.0            0.0   
average       5.0  5.0      5.0  ...       1.0        1.0            1.0   

storeID  이안인테리어  BBQ 청룡점  남도찌개마을  만만어사밀키트 신대방점  코바코 신림역점  이모네포차  참나무한방통닭구이  
userID                                                                      
vu7pe0      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vuvv60      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vv3ge2      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vv3ucg      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vv3vh1      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vvetso      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
vvm4sd      0.0      0.0     0.0           0.0       0.0    0.0        0.0  
average     1.0      1.0     1.0           1.0       1.0    1.0        1.0  

[8 rows x 2314 columns]

In [320]:
#피클로 데이터 프레임 저장 
ratings_matrix.to_pickle("ratings_matrix.pkl")
ratings_matrix_save = pd.read_pickle("ratings_matrix.pkl")
ratings_matrix_save.head(6)

storeID  #술비 부평점  상록수온누리약국  서울닭집  서울닭튀김  서울삼족탕  소양강민물장어  손문대구막창갈매기살 시립대점  \
userID                                                                     
1006bu2      0.0       0.0   0.0    0.0    0.0      0.0              0.0   
100dlpu      0.0       0.0   0.0    0.0    0.0      0.0              0.0   
100i22n      0.0       0.0   0.0    0.0    0.0      0.0              0.0   
100is6n      0.0       0.0   0.0    0.0    0.0      0.0              0.0   
100q04c      0.0       0.0   0.0    0.0    0.0      0.0              0.0   
100rl4h      0.0       0.0   0.0    0.0    0.0      0.0              0.0   

storeID  수학에미친사람들  술냥이  숯먹은생고기촌  ...  소소떡볶이 관악점  손큰원조할매순대국 미아점  이안인테리어  \
userID                           ...                                     
1006bu2       0.0  0.0      0.0  ...        0.0            0.0     0.0   
100dlpu       0.0  0.0      0.0  ...        0.0            0.0     0.0   
100i22n       0.0  0.0      0.0  ...        0.0            0.0     0.0   
100is6n       0.0  0.0      0.0  ...        0.0            0.0     0.0   
100q04c       0.0  0.0      0.0  ...        0.0            0.0     0.0   
100rl4h       0.0  0.0      0.0  ...        0.0            0.0     0.0   

storeID  BBQ 청룡점  남도찌개마을  만만어사밀키트 신대방점  코바코 신림역점  이모네포차  참나무한방통닭구이  \
userID                                                               
1006bu2      0.0     0.0           0.0       0.0    0.0        0.0   
100dlpu      0.0     0.0           0.0       0.0    0.0        0.0   
100i22n      0.0     0.0           0.0       0.0    0.0        0.0   
100is6n      0.0     0.0           0.0       0.0    0.0        0.0   
100q04c      0.0     0.0           0.0       0.0    0.0        0.0   
100rl4h      0.0     0.0           0.0       0.0    0.0        0.0   

storeID  Unnamed: 482  
userID                 
1006bu2           0.0  
100dlpu           0.0  
100i22n           0.0  
100is6n           0.0  
100q04c           0.0  
100rl4h           0.0  

[6 rows x 2315 columns]

### 아이템 간 유사도 산출; 
코사인 유사도 기반, 사이킷런의 cosine_similarity()를 사용해 측정

이 함수는 행을 기준으로 서로 다른 행을 비교해 유사도를 산출 

so, userId기준인 행 레벨 데이터이므로 지금 consine_similiarity를 적용하면 아이템이 아닌 사용자 간 유사도가 만들어짐

현재의 ratings_matirx가 행; 아이템, 열;사용자 가 되어야 하므로, 
행렬 위치 변경 ➡️ Pandas의 transpose()함수 사용 


In [232]:
ratings_matrix_T = ratings_matrix.transpose()
ratings_matrix_T.head(3)

userID        1006bu2  100dlpu  100i22n  100is6n  100q04c  100rl4h  1017ci9  \
storeID                                                                       
#술비 부평점           0.0      0.0      0.0      0.0      0.0      0.0      0.0   
1인고기도시락혼자고기야      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
24시시골집밥상          0.0      0.0      0.0      0.0      0.0      0.0      0.0   

userID        101inpj  101iuk0  101o0lr  ...  vr98fb   vrc2nf   vu551j   \
storeID                                  ...                              
#술비 부평점           0.0      0.0      0.0  ...      0.0      0.0      0.0   
1인고기도시락혼자고기야      0.0      0.0      0.0  ...      0.0      0.0      0.0   
24시시골집밥상          0.0      0.0      0.0  ...      0.0      0.0      0.0   

userID        vu7pe0   vuvv60   vv3ge2   vv3ucg   vv3vh1   vvetso   vvm4sd   
storeID                                                                      
#술비 부평점           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
1인고기도시락혼자고기야      0.0      0.0      0.0      0.0      0.0      0.0      0.0  
24시시골집밥상          0.0      0.0      0.0      0.0      0.0      0.0      0.0  

[3 rows x 17747 columns]

ratings_matrix를 전치 행렬 기반으로 변환한 데이터 세트를 기반으로 영화의 코사인 유사도 구하기 + 직관적인 매장 유사도 값을 표현하기 위해 Cosine_similarity()로 반환된 넘파이 행렬에 매장명을 매핑해 Dataframe으로 변환

In [233]:
#코사인 유사도 구하기
from sklearn.metrics.pairwise import cosine_similarity
item_sim = cosine_similarity(ratings_matrix_T, ratings_matrix_T)

#Cosine similarity로 반환된 넘파이 행렬을 매장명을 매핑해 Dataframe으로 변환
item_sim_df = pd.DataFrame(data=item_sim, 
                           index=ratings_matrix.columns, 
                           columns=ratings_matrix.columns)

print(item_sim_df.shape)
item_sim_df.head(3)

(2315, 2315)


storeID       #술비 부평점  1인고기도시락혼자고기야  24시시골집밥상  60계 서울신림난곡점  777카페  88옛날통닭 창동점  \
storeID                                                                         
#술비 부평점           1.0           0.0       0.0          0.0    0.0         0.0   
1인고기도시락혼자고기야      0.0           1.0       0.0          0.0    0.0         0.0   
24시시골집밥상          0.0           0.0       1.0          0.0    0.0         0.0   

storeID       BBQ 난곡점  BBQ 봉천우성점  BBQ 사당역점  BBQ 서울대글로벌공학관점  ...  \
storeID                                                     ...   
#술비 부평점           0.0        0.0       0.0             0.0  ...   
1인고기도시락혼자고기야      0.0        0.0       0.0             0.0  ...   
24시시골집밥상          0.0        0.0       0.0             0.0  ...   

storeID       함지박치즈등갈비 신림2호점  핫도리탕  햇빛식당  혼맛스시  홍마떡 신림점  황솔촌 충장점  흥부족발보쌈  \
storeID                                                                    
#술비 부평점                  0.0   0.0   0.0   0.0      0.0      0.0     0.0   
1인고기도시락혼자고기야             0.0   0.0   0.0   0.0      0.0      0.0     0.0   
24시시골집밥상                 0.0   0.0   0.0   0.0      0.0      0.0     0.0   

storeID       흥부찜닭 공수간 삼겹본능 신림점  희래등 호계본점  희래원  
storeID                                         
#술비 부평점                     0.0       0.0  0.0  
1인고기도시락혼자고기야                0.0       0.0  0.0  
24시시골집밥상                    0.0       0.0  0.0  

[3 rows x 2315 columns]

Godfather, The (1972), Inception (2010)과 유사도가 높은 상위 5개 영화 추출

In [234]:
item_sim_df["텐동요츠야"].sort_values(ascending=False)[1:6]

storeID
에그썸                 0.164812
더멜팅팟                0.122994
나인온스버거_x            0.116281
나인온스버거_y            0.115212
오늘와인한잔 서울대역1번출구점    0.112499
Name: 텐동요츠야, dtype: float64

### 아이템 유사도 데이터로 개인화된 영화 추천 알고리즘 만들기
위 유사도 데이터는 사용자 평점 기준이지만 매장 간 유사도만 가지고 매장을 추천한다. 
\+ 최근접 이웃 협업 필터링으로 개인에게 최적화된 매장 추천을 구현

➡️ 아직 방문X 매장에 대해서 아이템 유사도와 기존에 방문한 매장의 평점 데이터를 기반으로해 모든 영화의 예측 평점을 계산한 후 높은 예측 평점을 가진 영화를 추천

(설명추가 - p.620)
이때 N은 최근접 이웃 범위 계수(item neighbor)를 의미 ➡️ 특정 아이템과 유사도가 가장 높은 Top-N개의 다른 아이템을 추출하는 데 사용; 
먼저 N의 범위에 제약을 두지 않고 모든 아이템으로 가정하고 예측 평점을 구하는 로직을 작성한 뒤에, Top-N아이템을 기반으로 협업 필터링을 수행하는 로직으로 변경

---
영화 간 유사도를 가지는 Dataframe인 item_sim_df와 사용자-영화 평점 dataframe인 ratings_matrix변수를 활용해 사용자별로 최적화된 평점 스코어를 예측하는 함수 생성 
➡️ predict_rating()이며, 인자로 사용자-영화 평점 넘파이 행렬(Rating_matrix를 넘파이 행렬로 변환)과 영화 간 유사도를 가지는 넘파이 행렬(item_sim_df를 넘파이 행렬로 변환)을 입력받고 앞서 R^u,i 식으로 개인화된 예측 평점을 계산. 

---


In [235]:
#N의 범위에 제약을 두지 않는다면, 사용자별 영화 예측 평점 R_u,i는 사용자 u의 모든 영화에 대한 실제 평점과 
    #영화 i의 다른 모든 영화와의 코사인 유사도를 벡터 내적 곱(dot)한 값을 정규화를 위해 ***(p.620)로 나눈 것을 의미

def predict_rating(ratings_arr, item_sim_arr ):
    ratings_pred = ratings_arr.dot(item_sim_arr)/ np.array([np.abs(item_sim_arr).sum(axis=1)])
    return ratings_pred

In [236]:
ratings_pred = predict_rating(ratings_matrix.values, item_sim_df.values)
ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index=ratings_matrix.index, \
                                  columns = ratings_matrix.columns)
ratings_pred_matrix.head(3)

/var/folders/sb/ggzbw3nx66b6nbvcr0yybr2w0000gn/T/ipykernel_57966/1555214377.py:5: RuntimeWarning: invalid value encountered in true_divide
  ratings_pred = ratings_arr.dot(item_sim_arr)/ np.array([np.abs(item_sim_arr).sum(axis=1)])


storeID  #술비 부평점  1인고기도시락혼자고기야  24시시골집밥상  60계 서울신림난곡점     777카페  88옛날통닭 창동점  \
userID                                                                        
1006bu2      0.0           0.0       0.0          0.0  0.013906         0.0   
100dlpu      0.0           0.0       0.0          0.0  0.044287         0.0   
100i22n      0.0           0.0       0.0          0.0  0.053765         0.0   

storeID  BBQ 난곡점  BBQ 봉천우성점  BBQ 사당역점  BBQ 서울대글로벌공학관점  ...  함지박치즈등갈비 신림2호점  \
userID                                                 ...                   
1006bu2      0.0   0.000000       0.0        0.023934  ...        0.018138   
100dlpu      0.0   0.000000       0.0        0.050815  ...        0.024867   
100i22n      0.0   0.066086       0.0        0.142731  ...        0.050119   

storeID  핫도리탕  햇빛식당      혼맛스시  홍마떡 신림점   황솔촌 충장점    흥부족발보쌈  흥부찜닭 공수간 삼겹본능 신림점  \
userID                                                                          
1006bu2   0.0   0.0  0.020726      0.0  0.000000  0.000000                0.0   
100dlpu   0.0   0.0  0.000000      0.0  0.092549  0.112004                0.0   
100i22n   0.0   0.0  0.070715      0.0  0.046275  0.000000                0.0   

storeID  희래등 호계본점  희래원  
userID                  
1006bu2  0.000000  0.0  
100dlpu  0.071923  0.0  
100i22n  0.000000  0.0  

[3 rows x 2315 columns]

예측 평점이 실제 평점과 영화의 코사인 유사도를 내적(dot)한 값이기 때문에, 기존에 영화를 관람하지 않아 0에 해당했던 실제 영화 평점이 예측에서는 값이 부여되는 경우가 많이 발생한다. (예측 평점이 실 평점에 비해 작을 수 있음) 이는 내적 결과를 코사인 유사도 벡터 합으로 나누었기 때문에 생기는 현상

➡️ 예측 결과가 실제 평점과 얼마나 차이가 있는지 확인 
: 예측 평가 지표는 MSE(평균제곱오차)를 적용 

이때 유의점은, 사요자가 영화 평점을 주지 않은 경우 앞에서는 평점을 0으로 부과했지만, 앞에서 개인화된 예측 점수는 평점을 주지 않은 영화에 대해서도 아이템 유사도에 기반해 평점을 예측함. 따라서 여기서 실제와 예측 평점의 차이는 기존에 평점에 부여된 데이터에 대해서만 오차 정도를 측정

In [237]:
from sklearn.metrics import mean_squared_error

#사용자가 평점을 부여한 영화에 대해서만 예측 성능 평가 MSE를 구함 
def get_mse(pred, actual):
    #평점이 있는 영화만 추출
    pred = pred[actual.nonzero()].flatten()
    actual = actual[actual.nonzero()].flatten()
    return mean_squared_error(pred, actual)

print('아이템 기반 모든 최근접 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values))
### MSE 7.008449706348125

아이템 기반 모든 최근접 이웃 MSE:  7.008449706348125


MSE = 9.9: 실제값과 예측값은 서로 스케일이 다르기 때문에 MSE가 클 수도 있다. 중요한 것은 MSE를 감소시키는 방향으로 개선하는 것

➡️ 앞의 predict_rating( ) 함수는 사용자별 영화의 예측 평점을 계산하기 위해 해당 영화와 다른 모든 영화 간의 유사도 벡터를 적용한 것.
많은 영화의 유사도 벡터를 이용하다 보니 상대적으로 평점 예측이 떨어짐. 

➡️ 특정 영화와 가장 비슷한 유사도를 가지는 영화에 대해서만 유사도 벡터를 적용하는 함수로 변경 
: predict_rating_topsim(ratings_arr, item_sim_arr, N=20)함수는 N인자를 가지고 있어 TOP-N 유사도를 가지는 영화 유사도 벡터만 예측값을 계산하는 데 적용

*다만 개별 예측값을 구하기 위해 행렬별로 for loop 반복 수행; TOP-N 유사도 벡터를 계산해야 하기 때문에 수행시간이 오래 걸림.


In [238]:
def predict_rating_topsim(ratings_arr, item_sim_arr, n=20):
    #사용자-아이템 평점 행렬 크기만큼 0으로 채운 예측 행렬 초기화
    
    pred = np.zeros(ratings_arr.shape)
    
    #열 크기만큼 루프 수행
    for col in range(ratings_arr.shape[1]):
        
        #유사도 행렬에서 유사도가 큰 순으로 n개 데이터 행렬의 인덱스 반환
        top_n_items = [np.argsort(item_sim_arr[:, col])[:-n-1:-1]]
        
        #개인화된 예측 평점을 계산
        for row in range(ratings_arr.shape[0]):
            pred[row, col] = item_sim_arr[col, :][top_n_items].dot(ratings_arr[row,:][top_n_items].T)
            pred[row, col] /= np.sum(np.abs(item_sim_arr[col,:][top_n_items]))
    
    return pred

In [239]:
#predict_rating_topsim()함수로 예측 평점 계산, 실제 평점과의 MSE를 구하기 & 예측 평점 넘파이 행렬은 판다스 Dataframe으로 재생성
ratings_pred = predict_rating_topsim(ratings_matrix.values, item_sim_df.values, n=20)
print('아이템 기반 최근접 TOP-20 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values))

#예측 평점 데이터는 Dataframe으로 재생성 

ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index=ratings_matrix.index, columns=ratings_matrix.columns)

/var/folders/sb/ggzbw3nx66b6nbvcr0yybr2w0000gn/T/ipykernel_57966/1830049633.py:14: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  pred[row, col] = item_sim_arr[col, :][top_n_items].dot(ratings_arr[row,:][top_n_items].T)
/var/folders/sb/ggzbw3nx66b6nbvcr0yybr2w0000gn/T/ipykernel_57966/1830049633.py:15: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  pred[row, col] /= np.sum(np.abs(item_sim_arr[col,:][top_n_items]))
/var/folders/sb/ggzbw3nx66b6nbvcr0yybr2w0000gn/T/ipykernel_57966/1830049633.py:15: RuntimeWarning: invalid value encountered in d

아이템 기반 최근접 TOP-20 이웃 MSE:  4.123726002831022


➡️ 기존 9.89보다 향상됨(4.123726002831022)

### 특정 사용자에 대한 영화 추천; 
userId = '1006bu2'인 사용자가 어떤 영화에 평점을 높이 주었는지 확인

In [243]:
user_rating_id = ratings_matrix.locU['100i22n', :]
user_rating_id[ user_rating_id > 0].sort_values(ascending=False)[:10]

storeID
행운동조개            5.0
돈뼈락연탄갈비_x        4.0
돈뼈락연탄갈비_y        4.0
정숙성              4.0
퍼블리코타코           4.0
아궁이빙수            3.0
꼬꼬아찌숯불치킨 신림역점    3.0
스시공방             3.0
빠레뜨한남 샤로수길점      3.0
청기와타운 서울대점       3.0
Name: 100i22n, dtype: float64

이 사용자에게 아이템 기반 협업 필터링을 통해 영화를 추천 - 이미 평점을 준 영화를 제외하고 추천할 수 있도록, 평점을 주지 않은 영화를 리스트 객체로 반환하는 함수인 get_unseen_movies( )를 정의

In [241]:
def get_unseen_movies(ratings_matrix, userId):
    #userId로 입력받은 사용자의 모든 영화 정보를 추출해 Series로 반환
    #반환된 user_rating은 영화명(title)을 인덱스로 가지는 Series 객체
    user_rating = ratings_matrix.loc[userId, :]
    
    #user_rating이 0보다 크면 기존에 관람한 영화이다. 대상 인덱스를 추출해 list 객체로 만든다
    already_seen = user_rating[user_rating > 0].index.tolist()
    
    #모든 영화명을 list객체로 만듦
    movies_list = ratings_matrix.columns.tolist()
    
    #list comprehension으로 already_seen에 해당하는 영화는 movies_list에서 제외함
    unseen_list = [movie for movie in movies_list if movie not in already_seen]
    
    return unseen_list

사용자가 영화의 평점을 주지 않은 추천 대상 영화 정보와 predict_rating_topsim()에서 추출한 사용자별 아이템 유사도에 기반한 예측 평점 데이터셋을 이용해 최종적으로 사용자에게 영화를 추천하는 함수인 recomm_movie_by_userid( )를 정의

해당 함수는 예측 평점 DataFrame과 추천하려는 사용자 id, 추천 후보 영화 리스트, 추천 상위 영화 개수를 인자로 받아서 사용자가 좋아할만한 가장 높은 예측 평점을 가진 영화를 추천

In [244]:
def recomm_movie_by_userid(pred_df, userId, unseen_list, top_n=10):
    #예측 평점 Dataframe에서 사용자id 인덱스와 unseen_list로 들어온 매장명 칼럼을 추출
    #가장 예측 평점이 높은 순으로 정렬
    recomm_movies = pred_df.loc[userId, unseen_list].sort_values(ascending=False)[:top_n]
    return recomm_movies
    
#사용자가 방문하지 않은 매장명 추출
unseen_list = get_unseen_movies(ratings_matrix., '100i22n')

#아이템 기반 최근접 이웃 협업 필터링으로 생성
recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, '100i22n', unseen_list, top_n=10)
recomm_movies = pd.DataFrame(data = recomm_movies.values, index = recomm_movies.index, columns=['pred_score'])
recomm_movies

pred_score
storeID                
돈뼈락연탄갈비        1.431168
계림원 서울대입구역점    0.647642
소소떡볶이 관악점      0.646418
요리요리           0.559816
보울             0.489542
낙성대골목길         0.459062
원더트위스트         0.440597
삼차             0.439080
더멜팅팟           0.433341
남원추어탕_y        0.418736

In [352]:
ratings_matrix.drop(ratings_matrix.tail(1).index,inplace=True)
for i in ratings_matrix.index:

    unseen_list = get_unseen_movies(ratings_matrix, i)

    recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, i, unseen_list, top_n=10)
    recomm_movies = pd.DataFrame(data = recomm_movies.values, index = recomm_movies.index, columns=['pred_score'])
    saveit = i, recomm_movies
    saveit.to_pickle("ratings_matrix_all.pkl")

    
    
    
    
    
ratings_matrix_save = pd.read_pickle("ratings_matrix.pkl")
ratings_matrix_save.head(6)

AttributeError: 'tuple' object has no attribute 'to_pickle'